Ship-side part of the Mac fullscreen fix

larsy1995 opened the pull request:
New:
Adds an ifndef __APPLE__ to SohMenuSettings.cpp to remove the windowed mode checkbox from macOS since macOS is only going to use its native composited fullscreen-ing mode anyway, making this checkbox do nothing on that platform. 
OLD:
Changes to SohMenuBar.cpp needed to remove fullscreen mode toggle when in fullscreen and change the toggle text from windowed fullscreen to exclusive fullscreen.

Thanks @ReddestDream for making the fullscreen toggle fix. 

Dependent on:
https://github.com/Kenix3/libultraship/pull/817

<!--- section:artifacts:start -->
### Build Artifacts
  - [soh-linux.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/2830897678.zip)
  - [soh-mac.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/2830899477.zip)
  - [soh-windows.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/2830927590.zip)
<!--- section:artifacts:end -->

larsy1995 committed:
update

larsy1995 committed:
Revert "update"

This reverts commit 124687ad0a20f87e00ec2de71ac7b7118e1ce708.

larsy1995 committed:
Merge remote-tracking branch 'upstream/develop' into develop

larsy1995 committed:
Merge remote-tracking branch 'upstream/develop' into develop

larsy1995 committed:
Merge remote-tracking branch 'upstream/develop' into develop

larsy1995 committed:
Update libultraship

larsy1995 committed:
Update libultraship

larsy1995 committed:
Update libultraship

larsy1995 committed:
LUS Bump + Windowed -> Exclusive

larsy1995 committed:
Merge remote-tracking branch 'upstream/develop' into develop

larsy1995 committed:
Update libultraship

larsy1995 committed:
Merge remote-tracking branch 'upstream/develop' into develop

larsy1995 committed:
Here're the full changes for the fullscreen fix

larsy1995 committed:
Re-added a line

larsy1995 committed:
Without spaces this time

briaguya0 reviewed (commented):
after looking at this and the LUS PR i'm not quite following the logic here

what i understand:
* this changes the meaning of the windowed fullscreen cvar on mac
* this hides the toggle when in fullscreen on mac

i guess my main question is when would someone want to check the box vs not check the box? does just always using macos native fullscreen for fullscreen not work?

larsy1995 commented:
Exclusive mode enabled the direct rendering mode on macOS instead of just the composited rendering mode. The reason to keep the button at the moment is to not remove a feature that someone potentially might want to use, it is at least useful for testing and it keeps a sort of feature parity with the other operating systems.

But as it is currently, exclusive fullscreen is the default which is not ideal in a macOS context, thus I am inverting the CVAR on LUS and correcting the name on ship.

briaguya0 commented:
> someone potentially might want to use, it is at least useful for testing

i'm still not fully understanding *why* someone would want to use one or the other

> exclusive fullscreen is the default which is not ideal in a macOS context

why is this not ideal?

and my last question is about hiding the toggle when in fullscreen - that just feels like strange UX to me

larsy1995 commented:
We hide the toggle because otherwise we end up fullscreening twice if the fullscreen mode changes when in native fullscreen mode. This is a way to keep the behaviour consistent while keeping the mouse cursor hiding functionality when the imGUI menu is visible or hidden. 
Exclusive fullscreen blocks the macOS menubar and dock access as well as block out the notifications pop-up. This is not how most macOS applications handle this and this toggle is to make it more consistent with the macOS ecosystem as a whole. 
Removing exclusive fullscreen mode and just keeping native would be easier, but I was advised against removing functionality.

briaguya0 commented:
@Archez and/or @garrettjoecox probably have a better sense of the "why* for these things. my main hangup is inverting the cvar so on mac it means the opposite of what the value implies.

larsy1995 commented:
The other option would be to have the checkbox enabled by default, but I feel like that is a worse inconsistency.

briaguya0 commented:
> I was advised against removing functionality

who advised against that? did they give a reason? it sounds to me like just always using native fullscreen on mac is the logical choice

larsy1995 commented:
I can't remember, my memory doesn't keep information like that usually. 
I can definitely change the code to just use native fullscreen, that would simplify the code a fair bit, but what you're saying then is that we should not keep the direct rendering mode available and just permanently remove the toggle on macOS? 
I can do that.

briaguya0 commented:
i'd want to hear what @Archez and/or @garrettjoecox think about that idea first, but that feels reasonable to me

Malkierian commented:
Did I not have this handled with the new menu? If not, I feel like this should actually be part of the 9.0 milestone.

ReddestDream commented:
@Malkierian This one was not handled with the new menu (that was the scaling fix, not this fullscreen fix), but @larsy1995 is working on an updated version with @Archez AFAIK (https://github.com/Kenix3/libultraship/pull/817#issuecomment-2746428256). Thanks!

aMannus commented:
Yeah the reason I didn't get this in was because it looked like a completely different alternative was suggested.

larsy1995 commented:
Yeah, I’m hopefully going to update this PR today to make it ready for the updated fullscreen fix in LUS whenever that get’s added. Archez said he’d handle the LUS update for me, but I could also quickly make a new PR for that as well tbf.

larsy1995 committed:
Merge remote-tracking branch 'upstream/develop' into MacFullscreenFix

larsy1995 committed:
Made the windowed mode checkbox not appear on Mac

Since we're going to only use native macOS fullscreening anyway, there is no need to have a windowed mode checkbox as that wouldn't do anything.

Archez commented:
This will be handled instead by a LUS only change via https://github.com/Kenix3/libultraship/pull/855 which will remove Mac from `SupportsWindowedFullscreen()`, which the menu disabled attribute is obeying to hide the toggle.

larsy1995 commented:
Good good, I'll close this one then.
